[Goh Jun Yi] iP

Junyi00 opened the pull request:
# Arii
Just Another Duke Application

## Why Arii
Arii frees your mind of having to remember things you need to do!
- text-based
- quick to learn, easy to navigate
- ~~FAST~~ SUPER FAST to use

## Getting Started
Quick and easy setup steps, all you need to do is
1. Download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html)
2. Double-click it
3. Add your tasks
4. Let Arii work its magic!! 😉

## Features

- [X] Managing tasks
- [X] Search relevant tasks on a specific day
- [ ] Reminders (coming soon)

--- 
## Developers
If you are a Java programmer, you can use this to practice Java too! The `main` method:
```Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Add Gradle support

Junyi00 committed:
feat: Bot reads and echos back command

- Bot's name is now "Arii"
- Bot's behaviour is updated to greet, echo and exit

Junyi00 committed:
feat: Bot stores user texts and displays back when requested

- Stores user texts and displays them with the 'list' command

Junyi00 committed:
fix: Fix list starts from 1 instead

- Adjusted list displayed starts from index 1 instead of 0

Junyi00 committed:
feat: Add mark and unmark commands

- Mark and unmark commands have been added to change the status of a
  task

Junyi00 committed:
feat: Add new task types

- DeadlineTask, EventTask and TodoTask added
- deadline, event, todo commands have been added correspondingly

Junyi00 committed:
chore: Add missing space during display for EventTask

Junyi00 committed:
chore: Include testing of text ui

- Write tests for expected outcome of Duke given input

Junyi00 committed:
feat: Create DukeException class and add input validation

- DukeException class is created
- Input validation for bot commands

Junyi00 committed:
chore: Update text ui testing

- Update test for input validation

Junyi00 committed:
chore: More input validation for commands

Junyi00 committed:
chore: Update text ui testing for empty command details

Junyi00 committed:
feat: Add delete command to remove tasks

- New delete command to remove tasks created before

Junyi00 committed:
chore: Update text ui test for delete command

Junyi00 committed:
chore: Update to use enums instead

Junyi00 committed:
chore: Improve code styling

Junyi00 committed:
chore: Update javadocs

Junyi00 committed:
feat: Implement saving and loading of tasks

- One new command 'save', for saving of tasks to local storage
- Data is automatically loaded on intialisation

Junyi00 committed:
chore: Update .gitignore for data files

Junyi00 committed:
feat: Implement date parsing for task parameters

Junyi00 committed:
feat: Add command to check tasks on specific date

- New command 'check' that checks all tasks that are due to within given
  date

Junyi00 committed:
chore: Format displayed date

Junyi00 committed:
chore: Add javadoc for methods

Junyi00 committed:
chore: Include missing javadocs

Junyi00 committed:
Merge branch 'branch-level-8'

chore: Merge conflicts branch-level-8 to master

Junyi00 committed:
chore: Reorganise codebase into packages

Junyi00 committed:
chore: Add TaskList class

Junyi00 committed:
feat: Add classes Command, Ui, Parser and Storage

Junyi00 committed:
feat: Implement classes Command, TaskList, Parser, Storage and Ui

Junyi00 committed:
Merge branch 'add-gradle-support' of https://github.com/Junyi00/ip

Junyi00 committed:
Merge branch 'add-gradle-support'

Junyi00 committed:
chore: Update build.gradle

Junyi00 committed:
feat: Add JUnit tests

Junyi00 committed:
chore: Update javadocs

Junyi00 committed:
chore: Clean up import statements

Junyi00 committed:
feat: Add find command

Junyi00 committed:
Merge branch 'branch-A-CodingStandard'

Junyi00 committed:
Merge branch 'branch-Level-9'

Junyi00 committed:
feat: Initialise JavaFX dependencies and base UI

* Prepare for transition towards GUI-based application

chore: Initialise JavaFX dependencies and base UI

* Prepare for transition to GUI-based application

Junyi00 committed:
feat: Implement GUI for user interaction

Junyi00 committed:
chore: Include checkstyles and resolve checkstyle violations

Junyi00 committed:
Merge branch 'branch-A-CheckStyle'

chore: Merged with branch-A-CheckStyle and resolved conflicts

kohkaixun commented on src/main/java/duke/command/SaveCommand.java:
```diff
@@ -0,0 +1,42 @@
+ * A command to save all current tasks to the disk.
+ * @author Junyi
+ */
+public class SaveCommand extends Command {
+
+    private final TaskList taskList;
+    private final Storage storage;
+    private final Ui ui;
+
+    /**
+     * Constructor for SaveCommand.
+     * Saves the tasks to local storage.
+     * @param taskList TaskList of Duke's tasks.
+     * @param storage Storage instance of Duke.
+     * @param ui Ui instance of Duke.
+     */
+    public SaveCommand(TaskList taskList, Storage storage, Ui ui) {
+        this.taskList = taskList;
+        this.storage = storage;
+        this.ui = ui;
+    }
+
+    @Override
+    public String execute() throws DukeException {
+        try {
+            storage.save(taskList);
+            return ui.showSavedDataMessage();
+        } catch (IOException e) {
+            return "Arii can't access your files... Fix your system first.";
+        }
+    }
+}
```
If I'm not mistaken, it seems like you have implemented Arli in such a way where the user has to enter save before any of the changes made will be saved. For me, I just included this save method at the end of every method in TaskList that made a change to the TaskList, which was why I didn't have a SaveCommand class and if my Duke was to terminate by accident for some reason halfway through, the data inputted would still be saved. But I guess your way gives the user more control over what he can save and should the user decide that he or she isn't satisfied with what's been inputted, a simple restart of Arli will allow them to start over. Depends on what you want to emphasise on for Arli.

kohkaixun commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,198 @@
+        int taskIndex;
+        Command command;
+
+        try {
```
Nice! This is a much better way than how I implemented my parser which was basically a long chain of if else statements 😄

kohkaixun reviewed (commented):
LGTM. There wasn't really any issues with the coding standards so I just left an alternative way Arli's saving feature could work.

Junyi00 committed:
feat: Allow multiple keywords for find command

Junyi00 committed:
feat: Customise UI

* Use dark mode style instead
* Images of bot and user is replaced

Junyi00 committed:
chore: Include assertions and checks for empty user inputs

Currently users are able to send an empty command and the bot will reply
that something has went wrong.

Users may be confused by the returned dialog despite not sending an
actual command.

Validating the user input before executing the command allows us to
prevent such empty commands.

Validating the input before executing the user command is a
straightforward solution that does not complicate the Parser class
logic.

Junyi00 committed:
chore: Refactor code to follow SLAP

Currently, the parser class does not follow the SLAP principles due to
the validation logic.

This hurts the maintanability and readability of the Parser class as the
validation logic is occasionally abstracted while some are not. The
validation ocassionally comes in the middle of parsing which is not an
intuitive logic.

All validation logic are abstracted out and performed right at the start
when parsing the command. This gives a standardised order that is
easy to follow for readability.

Also, validation errors are now properly displayed to users due instead
of a generic message.

Junyi00 committed:
chore: Refactor to folow tell dont ask prinicple

Junyi00 committed:
feat: Sort tasks for easier viewing

Junyi00 committed:
Merge pull request #1 from Junyi00/branch-A-Assertions

Include assertions

Junyi00 committed:
Merge pull request #2 from Junyi00/branch-A-CodeQuality

Improve code quality

Junyi00 committed:
Merge branch 'master' into branch-C-Sort

Junyi00 committed:
Merge pull request #3 from Junyi00/branch-C-Sort

Sorted tasks

Junyi00 committed:
Create gradle.yml

Junyi00 committed:
chore: Resolve checkstyle issues

Junyi00 committed:
fix: Fix paths in runtest scripts

verydaftpunk commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,24 @@
+ * A command that asks Dukes to exit
+ * @author Junyi
+ */
+public class ByeCommand extends Command {
```
good using of inheritance

verydaftpunk commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,102 @@
+                String identifier = data.split(",")[0];
+                Task task;
+
+                switch (identifier) {
```
good following of coding standards and not having indentation for switch and case situation

verydaftpunk commented on src/main/java/duke/util/Ui.java:
```diff
@@ -0,0 +1,56 @@
+     * Displays a message that a task has been deleted.
+     */
+    public String showTaskDeletedMessage() {
+        return "Task deleted. Are you skipping on work again?";
```
good injection of personality into duke. very good.

verydaftpunk commented on .github/workflows/gradle.yml:
```diff
@@ -0,0 +1,50 @@
+name: Java CI
```
fantastic job!

Junyi00 committed:
feat: Improve GUI

* Fix truncation of text if bot reply is too long
* Bot reply now has a lighter background

Junyi00 committed:
chore: Give Arii more personality

Junyi00 committed:
chore: Add docs/Ui.png

Junyi00 committed:
chore: Update User Guide

Junyi00 committed:
chore: Improve code styling

* Fix long method in Parser.java
* Improved header comments

Junyi00 committed:
chore: Add more tests

Junyi00 committed:
fix: Fix broken bye command in GUI
